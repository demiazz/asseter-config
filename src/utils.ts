import Ajv from "ajv";
import { readFileSync } from "fs";

export const readJSON = (fileName: string): JSON => {
  const content = readFileSync(fileName, { encoding: "utf8" });

  return JSON.parse(content) as JSON;
};

export const readSchema = (fileName: string): Ajv.ValidateFunction => {
  const definition = readJSON(fileName);
  const schema = new Ajv({
    allErrors: true,
    jsonPointers: false,
  });

  return schema.compile(definition);
};

class ValidationError extends Error {
  constructor(errors: Ajv.ErrorObject[]) {
    super();

    this.message = errors.reduce<string[]>(
      (messages: string[], error: Ajv.ErrorObject) => {
        messages.push(`configuration${error.dataPath} ${error.message}\n`);

        return messages;
      },
      ["Asseter Invalid Configuration\n\n"]
    ).join('').trim();

    Error.captureStackTrace(this, this.constructor);
  }
}

export const validate = (
  schemaFileName: string,
  configurationFileName: string
): void => {
  const schema = readSchema(schemaFileName);
  const configuration = readJSON(configurationFileName);

  if (!schema(configuration)) {
    throw new ValidationError(schema.errors || []);
  }
};
