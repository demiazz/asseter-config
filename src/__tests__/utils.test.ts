import { readJSON, readSchema, validate } from "../utils";

import { getFixture, getFixturePath } from "./helpers";

describe("readJSON", () => {
  it("reads and parses JSON file", () => {
    const fileName = getFixturePath("helpers/readJSON.json");
    const expected = getFixture("helpers/readJSON.json");
    const actual = readJSON(fileName);

    expect(actual).toEqual(expected);
  });
});

describe("readSchema", () => {
  it("reads and parses JSON schema", () => {
    const schemaFileName = getFixturePath("helpers/schema.json");
    const schema = readSchema(schemaFileName);

    const valid = getFixture("helpers/valid.json");

    expect(schema(valid)).toBe(true);

    const invalid = getFixture("helpers/invalid.json");

    expect(schema(invalid)).toBe(false);
  });
});

describe("validate", () => {
  describe("when configuration is valid", () => {
    it("doesn't throw an error", () => {
      const schemaPath = getFixturePath("helpers/schema.json");
      const configurationPath = getFixturePath("helpers/valid.json");

      expect(() => {
        validate(schemaPath, configurationPath);
      }).not.toThrowError();
    });
  });

  describe("when configuration is invalid", () => {
    it("throws an error", () => {
      const schemaPath = getFixturePath("helpers/schema.json");
      const configurationPath = getFixturePath("helpers/invalid.json");

      expect(() => {
        validate(schemaPath, configurationPath);
      }).toThrowError();
    });

    it("throws an error with all errors information", () => {
      const schemaPath = getFixturePath("helpers/schema.json");
      const configurationPath = getFixturePath("helpers/invalid.json");

      expect(() => {
        validate(schemaPath, configurationPath);
      }).toThrowErrorMatchingSnapshot();
    });
  });
});
