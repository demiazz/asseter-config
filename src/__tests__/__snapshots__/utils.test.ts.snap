// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate when configuration is invalid throws an error with all errors information 1`] = `
"Asseter Invalid Configuration

configuration should NOT have additional properties
configuration.isValid should be equal to one of the allowed values"
`;
